File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/viewHolders/tasks/TodoViewHolder.java
package com.habitrpg.android.habitica.ui.viewHolders.tasks;

import android.view.View;

import com.habitrpg.android.habitica.models.tasks.Task;

import java.text.DateFormat;

public class TodoViewHolder extends ChecklistedViewHolder {

    private DateFormat dateFormatter;

    public TodoViewHolder(View itemView) {
        super(itemView);
        dateFormatter = android.text.format.DateFormat.getDateFormat(context);
    }

    @Override
    public void bindHolder(Task newTask, int position) {
        this.task = newTask;
        if (newTask.getCompleted()) {
            this.checklistIndicatorWrapper.setBackgroundColor(this.taskGray);
        } else {
            this.checklistIndicatorWrapper.setBackgroundResource(newTask.getLightTaskColor());
        }
        super.bindHolder(newTask, position);
    }

    @Override
    protected void configureSpecialTaskTextView(Task task) {
        if (this.specialTaskTextView != null) {
            if (task.getDueDate() != null) {
                this.specialTaskTextView.setText(dateFormatter.format(task.getDueDate()));
                this.specialTaskTextView.setVisibility(View.VISIBLE);
            } else {
                this.specialTaskTextView.setVisibility(View.INVISIBLE);
            }
        }
    }

    @Override
    public Boolean shouldDisplayAsActive(Task newTask) {
        return !newTask.getCompleted();
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/viewHolders/tasks/TodoViewHolder.kt
package com.habitrpg.android.habitica.ui.viewHolders.tasks

import android.view.View
import androidx.core.content.ContextCompat
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.extensions.formatForLocale
import com.habitrpg.android.habitica.helpers.GroupPlanInfoProvider
import com.habitrpg.android.habitica.models.tasks.ChecklistItem
import com.habitrpg.android.habitica.models.tasks.Task
import com.habitrpg.shared.habitica.models.responses.TaskDirection

class TodoViewHolder(
    itemView: View,
    scoreTaskFunc: ((Task, TaskDirection) -> Unit),
    scoreChecklistItemFunc: ((Task, ChecklistItem) -> Unit),
    openTaskFunc: ((Task, View) -> Unit),
    brokenTaskFunc: ((Task) -> Unit),
    assignedTextProvider: GroupPlanInfoProvider?,
) : ChecklistedViewHolder(itemView, scoreTaskFunc, scoreChecklistItemFunc, openTaskFunc, brokenTaskFunc, assignedTextProvider) {
    override fun bind(
        data: Task,
        position: Int,
        displayMode: String,
        ownerID: String?,
    ) {
        this.task = data
        setChecklistIndicatorBackgroundActive(data.isChecklistDisplayActive)
        reminderTextView.visibility = View.GONE
        this.streakTextView.visibility = View.GONE
        super.bind(data, position, displayMode, ownerID)
    }

    override fun configureSpecialTaskTextView(task: Task) {
        super.configureSpecialTaskTextView(task)
        if (task.dueDate != null) {
            if (task.isDueToday() == true) {
                specialTaskTextView?.text = context.getString(R.string.today)
            } else if (task.isDayOrMorePastDue() == true) {
                task.dueDate?.let { specialTaskTextView?.text = it.formatForLocale() }
                specialTaskTextView?.setTextColor(ContextCompat.getColor(context, R.color.maroon100_red100))
            } else {
                task.dueDate?.let { specialTaskTextView?.text = it.formatForLocale() }
                specialTaskTextView?.setTextColor(ContextCompat.getColor(context, R.color.gray_300))
            }
            this.specialTaskTextView?.visibility = View.VISIBLE
            calendarIconView?.visibility = View.VISIBLE
        } else {
            this.specialTaskTextView?.visibility = View.INVISIBLE
        }
    }

    override fun shouldDisplayAsActive(
        task: Task?,
        userID: String?,
    ): Boolean {
        return task?.completed(userID) != true
    }
}
